ShaiRandom.UnitTests/KnownSeriesRandomTests.cs

```diff
@@ -0,0 +1,365 @@
+﻿using System;
+using System.Diagnostics.CodeAnalysis;
+using System.Reflection;
+using ShaiRandom.Generators;
+using Xunit;
+
+namespace ShaiRandom.UnitTests
+{
+    [SuppressMessage("ReSharper", "UselessBinaryOperation")]
+    public class KnownSeriesRandomTests
+    {
+        private const int ReturnedValue = 10;
+
+        private static readonly float s_floatCloseTo1 = 1.0f - MathF.Pow(2f, -24f);
+        private static readonly double s_doubleCloseTo1 = 1.0 - Math.Pow(2, -53);
+
+        private readonly KnownSeriesRandom _boundedRNG = new KnownSeriesRandom(
+            new []{ReturnedValue},
+            new []{(uint)ReturnedValue},
+            new []{(double)ReturnedValue},
+            byteSeries: new []{(byte)ReturnedValue},
+            floatSeries: new []{(float)ReturnedValue},
+            longSeries: new []{(long)ReturnedValue},
+            ulongSeries: new []{(ulong)ReturnedValue}
+        );
+
+        private readonly KnownSeriesRandom _zeroRNG = new KnownSeriesRandom(
+            new []{0},
+            new []{(uint)0},
+            new []{(double)0},
+            byteSeries: new []{(byte)0},
+            floatSeries: new []{(float)0},
+            longSeries: new []{(long)0},
+            ulongSeries: new []{(ulong)0}
+        );
+
+        private readonly KnownSeriesRandom _unboundedRNG = new KnownSeriesRandom(
+            new []{int.MinValue, int.MaxValue},
+            new []{uint.MinValue, uint.MaxValue},
+            new []{0.0, s_doubleCloseTo1, -double.Epsilon, 1.0},
+            byteSeries: new []{byte.MinValue, byte.MaxValue},
+            floatSeries: new []{0.0f, s_floatCloseTo1, -float.Epsilon, 1.0f},
+            longSeries: new []{long.MinValue, long.MaxValue},
+            ulongSeries: new []{ulong.MinValue, ulong.MaxValue}
+        );
+
+        private readonly KnownSeriesRandom _unboundedInclusiveRNG = new KnownSeriesRandom(
+            doubleSeries: new []{0.0, 1.0, -double.Epsilon, 1.1},
+            floatSeries: new []{0.0f, 1.0f, -float.Epsilon, 1.1f}
+        );
+
+        private readonly KnownSeriesRandom _unboundedExclusiveRNG = new KnownSeriesRandom(
+            doubleSeries: new []{double.Epsilon, s_doubleCloseTo1, 0.0, 1.0},
+            floatSeries: new []{float.Epsilon, s_floatCloseTo1, 0.0f, 1.0f}
+        );
+        #region Template Tests
+
+        private (Func<T> unbounded, Func<T, T> outerBound, Func<T, T> zeroOuterBound, Func<T, T, T> dualBound) GetGenerationFunctions<T>(string name, KnownSeriesRandom unboundedRNG)
+            where T : notnull
```

Comment:
Woah, that's quite a signature. With the reflection below this as well, I can't gauge whether this is right by looking at it, but you've gotten code almost this complex fixed before, and I think it's reasonable to have faith in this code working.

This is still impressive in a scary sort of way.